src/linker/Linker.Dataflow/ValueNode.cs

```diff
@@ -1239,23 +1252,36 @@ public override bool Equals (ValueNode other)
 				return false;
 
 			ArrayValue otherArr = (ArrayValue) other;
-			return Size.Equals (otherArr.Size);
+			bool equals = Size.Equals (otherArr.Size);
+			equals &= IndexValues.Count == otherArr.IndexValues.Count;
+			if (!equals)
+				return false;
+
+			// If both sets T and O are the same size and "T intersect O" is empty, then T == O.
+			HashSet<KeyValuePair<int, ValueBasicBlockPair>> thisValueSet = new (IndexValues);
+			HashSet<KeyValuePair<int, ValueBasicBlockPair>> otherValueSet = new (otherArr.IndexValues);
+			thisValueSet.ExceptWith (otherValueSet);
+			return thisValueSet.Count == 0;
 		}
 
 		protected override string NodeToString ()
 		{
-			return ValueNodeDump.ValueNodeToString (this, Size);
+			// TODO: Use StringBuilder and remove Linq usage.
+			return $"(Array Size:{ValueNodeDump.ValueNodeToString (this, Size)}, Values:({string.Join (',', IndexValues.Select (v => $"({v.Key},{ValueNodeDump.ValueNodeToString (v.Value.Value)})"))})";
 		}
 
 		protected override IEnumerable<ValueNode> EvaluateUniqueValues ()
 		{
 			foreach (var sizeConst in Size.UniqueValuesInternal)
-				yield return new ArrayValue (sizeConst);
+				yield return new ArrayValue (sizeConst, ElementType, IndexValues);
 		}
 
 		protected override ValueNode ChildAt (int index)
 		{
 			if (index == 0) return Size;
+			if (index - 1 <= IndexValues.Count)
+				return IndexValues.Values.ElementAt (index - 1).Value;
+
 			throw new InvalidOperationException ();
```

Comment:
Sorry if I missed it - what was the resolution of my question? I don't think we should fail hard if we can avoid it.

Reply:
I realized that there was a bug in the ChildAt function for fetching from the IndexValues dictionary. The idea I had for what the "children" collection is for the ArrayValue object is to have each `index-1` element be the i-th element of the Values collection and to match other nodes with children, to throw an `InvalidOperationException` when an out of range child is requested. To fetch a value at a particular index, the user should use the `IndexValues` property instead.